### PR TITLE
Huggingface trainer and huge refactoring of models and pretty much everything, lol

### PR DESCRIPTION
**Pull request by sordonia**
- refactor of expert models such that they are disentangled with lightning logic
- keeps both lightning trainer and the hf trainer
- renamed config.py to arguments.py, in the next PR will rename ExpertConfig to ExpertArgs
- returns CausalLMOutput from ExpertModel, to align with HF stuff
- forward now does not accept a `batch` argument, to align with how HF expects models to behave
- added Serializable class that handles deserialization and Serialization of dataclasses by dumping the module path under mttl (this could break if we move modules around, so we might find a better solution afterwards, I also implemented dynamic loading, where the stored class is searched in the current loaded modules if the module stored in the serialized form is not found anymore)
- added HF training script and DownstreamEvalCallback for HF

### CODE REVIEW

**Review by pclucas14 — APPROVED** [review]
PR looks good. I think we'll have to play with it and see what happens :)

TO THE MOOOONNN  🚀  🚀  🚀

File: projects/modular_llm_hf/configs/models/mistral_7b.json
Comment (oleksost) [inline projects/modular_llm_hf/configs/models/mistral_7b.json:1]: can we use configs directly from modular_llm here?

File: mttl/models/hf/trainer.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            output_dir,
+            state_dict=state_dict,
+            safe_serialization=self.args.save_safetensors,
+        )
+
+        if self.tokenizer is not None:
+            self.tokenizer.save_pretrained(output_dir)
+
+        # Good practice: save your training arguments together with the trained model
+        torch.save(self.args, os.path.join(output_dir, TRAINING_ARGS_NAME))
```
Comment (oleksost) [inline mttl/models/hf/trainer.py:75]: Why do we need self.args and self.mttl_args?

Comment (oleksost) [reply mttl/models/hf/trainer.py:75]: is one containing the other?

File: mttl/models/hf/trainer.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+from mttl.models.get_optimizer import get_optimizer, get_optimizer_and_scheduler
+from mttl.models.get_scheduler import get_scheduler
+
+MTTL_ARGS_NAME = "mttl_args.bin"
+
+
+class ExpertModelTrainer(Trainer):
+    """Generic HF trainer for expert models."""
+
+    def __init__(self, model, args, **kwargs):
```
Comment (oleksost) [inline mttl/models/hf/trainer.py:21]:
```suggestion
    def __init__(self, model, args:TrainingArgs, **kwargs):
```

File: mttl/arguments.py [hunk 3]
```diff
@@ -454,6 +460,35 @@ def __post_init__(self):
...
+    def to_hf_training_args(self) -> "TrainingArguments":
+        from transformers import TrainingArguments
+
+        return TrainingArguments(
+            run_name=self.wandb_run_name or self.expert_name or self.finetune_task_name,
+            use_cpu=self.compute_strategy == "cpu",
+            overwrite_output_dir=True,
+            output_dir=self.output_dir,
+            per_device_train_batch_size=self.train_batch_size,
+            per_device_eval_batch_size=self.predict_batch_size,
```
Comment (oleksost) [inline mttl/arguments.py:472]: hm, `predict_batch_size` is not defined in this file. 

Sorry but idk now where`self.predict_batch_size` is it set now and why is it in `self`.

**Review by oleksost — APPROVED** [review]
Great PR, would need to play more with it indeed to figure out all the nitty details.

File: mttl/models/expert_model.py [hunk 4]
```diff
@@ -585,115 +388,98 @@ def save_to_library(self, library_id):
...
-        self.log(f"{self._log_pref}train/loss", loss, on_step=True, prog_bar=True)
-        self.log(
-            f"{self._log_pref}train/total_loss", total_loss, on_step=True, prog_bar=True
-        )
+@dataclass
+class MoEModelConfig(BaseExpertModelConfig):
+    library_id: str = None
+    moe_num_experts: int = 1
+    selector_config: AutoSelectorConfig = None
+    modifier_config: AutoModifierConfig = None
```
[the remark this answers is not in the document]
Comment (sordonia) [reply mttl/models/expert_model.py:473]: it's not?

File: projects/modular_llm_hf/configs/models/mistral_7b.json
Comment (sordonia) [reply projects/modular_llm_hf/configs/models/mistral_7b.json:1]: what do you mean?

File: mttl/models/hf/trainer.py (hunk 1 above)
Comment (sordonia) [reply mttl/models/hf/trainer.py:75]: self.args is the huggingface TrainingArguments
self.mttl_args is all the args for mttl

File: mttl/arguments.py (hunk 3 above)
Comment (sordonia) [reply mttl/arguments.py:472]: predict_batch_size and train_batch_size is automatically injected by the DatasetConfig

File: mttl/models/expert_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleksost) [reply mttl/models/expert_model.py:473]: I mean to suggest removing the line, cause its in the super class I think

File: projects/modular_llm_hf/configs/models/mistral_7b.json
Comment (oleksost) [reply projects/modular_llm_hf/configs/models/mistral_7b.json:1]: can we just e-used configs from the `modular_llm/configs` folder in the `modular_llm_hf`?

File: mttl/arguments.py (hunk 3 above)
Comment (oleksost) [reply mttl/arguments.py:472]: these automatic injections are confusion, idk. If we can do something about it here...

File: mttl/evaluators/base.py [hunk 5]
```diff
@@ -293,9 +298,9 @@ def generate_for_batch(self, model, batch):
         batch = transfer_batch_to_device(batch, device)
 
         with torch.no_grad():
-            if isinstance(model, EfficientCheckpointModule):
+            if isinstance(model, ExpertModule) or isinstance(model, ExpertModel):
                 predictions = model.generate(
-                    batch,
+                    **batch,
```
Comment (shuishen112) [inline mttl/evaluators/base.py:303]: nice!

**Review by shuishen112 — APPROVED** [review]

File: mttl/models/llama_patch.py [hunk 6]
```diff
@@ -13,12 +13,6 @@
 )
 
 try:
-    from einops import rearrange
```
Comment (matheper) [inline mttl/models/llama_patch.py:16]: Are we dropping flash_attn?

**Pull request merged**